Changes in PR #18665 will break exported native modules

- [x] I tried using the `@types/react-native` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @alloy @gyzerok @huhuanming @iRoachie @timwangdev @kamal

A recently merged PR #18665 will break for us since exported native modules rely on `NativeModules` to be of type `any` otherwise the compiler will complain about missing properties:

```
error TS2339: Property 'MyNativeModule' does not exist on type 'NativeModulesStatic'.
error TS2339: Property 'SettingsManager' does not exist on type 'NativeModulesStatic'.
error TS2339: Property 'I18nManager' does not exist on type 'NativeModulesStatic'.
```

--- a/types/react-native/index.d.ts
+++ b/types/react-native/index.d.ts
@@ -9126,9 +9126,12 @@ export var NativeEventEmitter: NativeEventEmitter
 export var NativeAppEventEmitter: RCTNativeAppEventEmitter
 
 /**
- * Empty interface which can be augmented by other type definitions for the NativeModules var below.
+ * Interface for NativeModules which allows to augment NativeModules with type informations.
+ * See react-native-sensor-manager for example.
  */
-interface NativeModulesStatic {}
+interface NativeModulesStatic {
+    [name: string]: any;
+}
 
 /**
  * Native Modules written in ObjectiveC/Swift/Java exposed via the RCTBridge

--- a/types/react-native/test/index.tsx
+++ b/types/react-native/test/index.tsx
@@ -39,7 +39,23 @@ import {
     ScrollViewProps,
     RefreshControl,
     TabBarIOS,
+    NativeModules
 } from 'react-native';
+
+declare module 'react-native' {
+    interface NativeTypedModule {
+        someFunction(): void;
+        someProperty: string;
+    }
+    interface NativeModulesStatic {
+        NativeTypedModule: NativeTypedModule
+    }
+}
+
+NativeModules.NativeUntypedModule;
+
+NativeModules.NativeTypedModule.someFunction();
+NativeModules.NativeTypedModule.someProperty = "";
 
 function testDimensions() {
   const {

